File: spherocontrol/cvtest.cpp
#include <iostream>
#include <vector>

#include <opencv2/opencv.hpp>

cv::Point2f warpPerspective(const cv::Point2f& p, const cv::Mat& m)
{
    cv::Point2f q(p.x * m.at<double>(0, 0) + p.y * m.at<double>(0, 1) + m.at<double>(0, 2),
        p.x * m.at<double>(1, 0) + p.y * m.at<double>(1, 1) + m.at<double>(1, 2));

    const double f = p.x * m.at<double>(2, 0) + p.y * m.at<double>(2, 1) + m.at<double>(2, 2);


    q.x /= f;
    q.y /= f;

    return q;
}

int main()
{
    float width = 5.0f, height = 5.0f;

    // std::vector<cv::Point2f> corners { { 0.0f, 0.0f }, { 4.0f, 1.0f }, { 3.4f, 15.0f }, { -0.5f, 13.7f } };
    std::vector<cv::Point2f> corners { { 0.0f, 0.0f }, { 4.0f, 0.0f }, { 4.0f, 4.0f }, { 0.0f, 4.0f } };
    std::vector<cv::Point2f> target { { 0, 0 }, { width, 0 }, { width, height }, { 0, height } };

    auto transform = cv::getPerspectiveTransform(corners, target);

    std::cout << transform << std::endl;

    cv::Point2f p(3.5f, 1.765f);

    p = warpPerspective(p, transform);

    std::cout << "\n" << p << std::endl;

    return 0;
}

File: spherocontrol/cvtest.js
"use strict";

var cv = require("opencv");
var Matrix = cv.Matrix;
var width = 5;
var height = 5.7;

var corners = [0, 0, 4, 1, 3.4, 15, -0.5, 13.7];
var target = [0, 0, width, 0, width, height, 0, height];

var transform = new Matrix.Eye(3, 3).getPerspectiveTransform(corners, target);

console.log(transform.row(0));
console.log(transform.row(1));
console.log(transform.row(2));
console.log();

var points = [3.5, 1.765, 1];
// var out = Matrix.Multiply(transform, points); //transform.Multiply(points);
console.log(JSON.stringify(transform, null, '    '));

// var cv = require('opencv');
//
// cv.readImage("./mpi.png", function(err, im) {
//   if (err) throw err;
//
//   var width = im.width();
//   var height = im.height();
//   if (width < 1 || height < 1) throw new Error('Image has no size');
//
//   var srcArray = [0, 0, width, 0, width, height, 0, height];
//   var dstArray = [0, 0, width * 0.9, height * 0.1, width, height, width * 0.2, height * 0.8];
//   var xfrmMat = im.getPerspectiveTransform(srcArray, dstArray);
//   console.log(xfrmMat.row(0));
//   console.log(xfrmMat.row(1));
//   console.log(xfrmMat.row(2));
//   im.warpPerspective(xfrmMat, width, height, [255, 255, 255]);
//   im.save("./warp-image.png");
//   console.log('Image saved to ./tmp/warp-image.png');
// });
